### PR TITLE
8293824: gc/whitebox/TestConcMarkCycleWB.java failed "RuntimeException: assertTrue: expected true, was false"

### PR DESCRIPTION
**Pull request by kimbarrett**
Please review this change to WhiteBox and some tests involving G1 concurrent GCs.

Some tests currently use WhiteBox.g1StartConcMarkCycle() to trigger a
concurrent GC.  Many of them follow it with a loop waiting for a concurrent
cycle to not be in progress.  A few also preceed that call with a similar
loop, since that call does nothing and returns false if a concurrent cycle is
already in progress.  Those tests typically want to ensure there was a
concurrent cycle that was started after some setup.

The failing test calls that function, asserting that it returned true, e.g. a
new concurrent cycle was started.

There are various problems with this, due to races with concurrent cycles
started automatically and possibly aborted (by full GCs) concurrent cycles,
making some of these tests unreliable in some configurations.

For example, the test failure associated with this bug intermittently arises
when running with `-Xcomp`, triggering a concurrent cycle before the explicit
requests by the test, causing the explicit request to fail (because there is
already one in progress), failing the assert.  Waiting for there not to be an
in-progress cycle before the explicit request just narrows the race window.

We have a different mechanism for controlling concurrent cycles, the
concurrent GC breakpoint mechanism.  By adding a counter specifically for such
cycles, we can use GC breakpoints to ensure only the concurrent cycles the
test wants are occurring, and can verify they completed successfully.

So we change tests using WhiteBox.g1StartConcMarkCycle() to instead use GC
breakpoints, along with the new WhiteBox.g1CompletedConcurrentMarkCycles() to
avoid racing request problems and to detect aborted cycles.  Since it is no
longer used, WhiteBox.g1StartConcMarkCycle() is removed.

Testing:
mach5 tier1-6

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8293824](https://bugs.openjdk.org/browse/JDK-8293824): gc/whitebox/TestConcMarkCycleWB.java failed "RuntimeException: assertTrue: expected true, was false"

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11435/head:pull/11435` \
`$ git checkout pull/11435`

Update a local copy of the PR: \
`$ git checkout pull/11435` \
`$ git pull https://git.openjdk.org/jdk pull/11435/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11435`

View PR using the GUI difftool: \
`$ git pr show -t 11435`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11435.diff">https://git.openjdk.org/jdk/pull/11435.diff</a>

</details>

### CODE REVIEW

File: test/hotspot/jtreg/gc/stress/TestStressRSetCoarsening.java [hunk 1]
```diff
@@ -293,10 +293,7 @@ public void go() throws InterruptedException {
                 if (pre > cur) {
                     // Number of references went down.
                     // Need to provoke recalculation of RSet.
-                    WB.g1StartConcMarkCycle();
-                    while (WB.g1InConcurrentMark()) {
-                        Thread.sleep(1);
-                    }
+                    WB.g1RunConcurrentGC(false);
```
Comment (walulyai) [inline test/hotspot/jtreg/gc/stress/TestStressRSetCoarsening.java:296]: probably use `WB.g1RunConcurrentGC();` overload for consistency with other calls

File: test/hotspot/jtreg/gc/g1/humongousObjects/objectGraphTest/GCTokens.java [hunk 2]
```diff
@@ -33,7 +33,7 @@ private GCTokens() {
 
     public static final String WB_INITIATED_YOUNG_GC = "Young (Normal) (WhiteBox Initiated Young GC)";
     public static final String WB_INITIATED_MIXED_GC = "Young (Mixed) (WhiteBox Initiated Young GC)";
-    public static final String WB_INITIATED_CMC = "WhiteBox Initiated Concurrent Mark";
+    public static final String WB_INITIATED_CMC = "WhiteBox Initiated Run to Breakpoint";
```
Comment (walulyai) [inline test/hotspot/jtreg/gc/g1/humongousObjects/objectGraphTest/GCTokens.java:36]: `WB_INITIATED_CMC = "WhiteBox Initiated Run to Concurrent GC Breakpoint"; ` as we indicate below that `String CMC = "Concurrent Mark)"`

**Review by walulyai — APPROVED** [review]
Lgtm!

Minor comments:

File: test/hotspot/jtreg/gc/g1/humongousObjects/objectGraphTest/GCTokens.java (hunk 2 above)
Comment (kimbarrett) [reply test/hotspot/jtreg/gc/g1/humongousObjects/objectGraphTest/GCTokens.java:36]: Sorry, I don't understand the question/comment?  The change to the string
reflects what is logged by the VM now that we're using GC breakpoints to
control the collection.  The logging being matched with CMC didn't change.

Comment (walulyai) [reply test/hotspot/jtreg/gc/g1/humongousObjects/objectGraphTest/GCTokens.java:36]: ok, then it is fine.

File: test/lib/jdk/test/whitebox/WhiteBox.java [hunk 3]
```diff
@@ -165,7 +165,52 @@ public         void addToSystemClassLoaderSearch(String segment) {
...
+
   public native boolean g1InConcurrentMark();
+  public native int g1CompletedConcurrentMarkCycles();
+
+  // Perform a complete concurrent GC cycle, using concurrent GC breakpoints.
+  // Completes any in-progress cycle before performing the requested cycle.
+  // Returns true if the cycle completed successfully.  If the cycle was not
+  // successful (e.g. it was aborted), then throws RuntimeException if
+  // errorIfFail is true, returning false otherwise.
+  public boolean g1RunConcurrentGC(boolean errorIfFail) {
```
Comment (walulyai) [inline test/lib/jdk/test/whitebox/WhiteBox.java:177]: Failure detection as described in the comment above is turned off for all test cases. Please verify if that is intentional.

**Review by walulyai — CHANGES_REQUESTED** [review]

File: test/lib/jdk/test/whitebox/WhiteBox.java (hunk 3 above)
Comment (kimbarrett) [reply test/lib/jdk/test/whitebox/WhiteBox.java:177]: Drat!  That's NOT intentional.  Fixing and retesting now.  Thanks for spotting it.

**Comment by kimbarrett** [comment]
I'm going to need to retarget this to jdk20.  Also, this change makes JDK-8298215 a reliable failure instead of very intermittent.  So closing this PR and I'll create a new one later, after doing something about JDK-8298215.

**Pull request closed without merging**